### ISSUE TITLE
Allow custom startup commands

### ISSUE THREAD

**Issue by anthonychu** [Azure/static-web-apps-cli#147]
**Is your feature request related to a problem? Please describe.**

When starting the emulator with a dev server, we need to start the dev server separately before starting `swa`.

**Describe the solution you'd like**

Allow an app startup command and working folder to be provided. The emulator starts the command before proceeding. For example, running a startup command in the frontend folder might look like this:

```
swa start http://localhost:3000 --app-startup-command "npm start" --app-location ./frontend
```

Or if we can make this work cross platform:

```
swa start http://localhost:3000 --app-startup-command "cd frontend && npm start"
```

Perhaps also support `--api-startup-command`? Not sure if this is needed right now.

**Describe alternatives you've considered**

Start the process separately.

**Additional context**

Also needs this functionality as we have to wait for the servers to be up before proceeding: #146

**Comment by manekinekko** [comment]
I think this is a duplicate of https://github.com/Azure/static-web-apps-cli/issues/144 ?

**Comment by anthonychu** [comment]
Thanks. Didn't know you already implemented it!

I tried it and can't figure out how it works.

What is the equivalent of `swa start http://localhost:3000 --app-startup-command "cd frontend && npm start"` above? I tried `swa start http://localhost:3000 --run "npm:start" --app frontend` and other variations.

How does the `npm:start` syntax work? Can I just pass any command (e.g., `dotnet watch run`)?

**Comment by manekinekko** [comment]
You can pass it a shell script to execute. This way we are not limiting the user to a single command. And also this avoid dealing with special character.

However, the equivalent of your command would be:

```
swa start http://localhost:3000 --run npm:start --app-location frontend
```

_(however, I haven't yet implemented #146 so this command may fail because port 3000 is not yet ready)_

> How does the npm:start syntax work?

`npm:start` follows this pattern `<npm or yarn>:<script in package.json>`

>  Can I just pass any command (e.g., dotnet watch run)?

You would need to provide a script and run `swa start http://localhost:3000 --run my-script.sh` – or any executable file. We simply run whatever file you provide here.

I included a [set of unit tests](https://github.com/Azure/static-web-apps-cli/blob/main/src/core/utils.test.ts#L791-L914) to illustrate what's valid and what's not.

**Comment by anthonychu** [comment]
I think a shell script is too heavy. How big of a problem is it with special characters? Can we also supporting taking a string that's the command?

I like the parameter name `run`, however I wonder if we ever need to specify how to run the function app too.

**Comment by manekinekko** [comment]
How about we simply allow passing custom commands like the one we have in the workflow file:
- `--app-build-command` // run from the --app-location folder
- `--api-build-command` // run from the --api-location folder

These flags would override what users have in their workflow file (if they have one).

Users can then pass in any command they want. Would that work?

**Comment by anthonychu** [comment]
I think there’s a different between “build” and “run”/“startup command”. Build implies run to completion, then start emulator. The command I’m looking for is run and start emulator when port is available. So I think this would work better as something other than a build command.

**Comment by manekinekko** [comment]
Would this work for you?

```
--run "cd frontend && npm start"
```

**Comment by anthonychu** [comment]
```
# option 1
--run "cd frontend && npm start"
```

This looks great!

Can you also provide more info on how `--run` works currently? I try it and nothing happens. For example, in a folder where I can run `npm run start`. If I run this command it gets stuck:

```
swa start http://localhost:3000 --run "npm:start"
⠼ Waiting for http://localhost:3000 to be ready...
```

**Comment by manekinekko** [comment]
Good catch! What's happening here is that when we start with a dev server context (`swa start http://localhost:3000 ...`), we await ([waitOn](https://www.npmjs.com/package/wait-on)) for that dev server to be ready before moving on. 

This process will block the rest of the script and hence we don't reach the `--run` logic.

To fix this, I suggest that we skip awaiting for the dev server if we are also passing in a `--run` command. This way we:
1. start swa and setup the proxy requests for the dev server 
2. in parallel, we process the `--run` command (which for instance starts the dev server)

However, if we follow this logic, then  running the following command `swa start http://localhost:3000 --run "echo Hey"`, where the `--run` script doesn't start a dev server, then all proxied requests to `http://localhost:3000` will fail if the dev server isn't started. (and we skipped checking if the dev server is ready).


On the other hand, if we don't passe in a `--run` script, then we will have to await for the dev server to be ready (what we are doing rn).
